File: frontend/src/utils/clustering.js
/**
 * HyperCrop — Clustering Utilities (Frontend)
 *
 * Prototype cluster detection and risk classification.
 * Replace with DBSCAN/HDBSCAN for production.
 */

import { haversineDistance } from './evidenceScore';

/**
 * Classify cluster risk using evidence-weighted logic.
 * NOT simply: 10 reports = high risk.
 */
export function classifyRisk(independentReports, confirmedReports, totalReports, avgEvidenceScore) {
  if (independentReports >= 6 && confirmedReports >= 4 && avgEvidenceScore > 0.55) return 'high';
  if (independentReports >= 4 && avgEvidenceScore > 0.45) return 'elevated';
  if (independentReports >= 2 || totalReports >= 5) return 'watch';
  return 'low';
}

/**
 * Get risk level metadata for display.
 */
export function getRiskMeta(riskLevel) {
  const meta = {
    high: {
      label: 'HIGH',
      color: '#DC2626',
      bgColor: 'rgba(220, 38, 38, 0.1)',
      borderColor: 'rgba(220, 38, 38, 0.3)',
      mapColor: '#DC2626',
      mapOpacity: 0.20,
    },
    elevated: {
      label: 'ELEVATED',
      color: '#EA580C',
      bgColor: 'rgba(234, 88, 12, 0.1)',
      borderColor: 'rgba(234, 88, 12, 0.3)',
      mapColor: '#EA580C',
      mapOpacity: 0.15,
    },
    watch: {
      label: 'WATCH',
      color: '#D97706',
      bgColor: 'rgba(217, 119, 6, 0.1)',
      borderColor: 'rgba(217, 119, 6, 0.3)',
      mapColor: '#D97706',
      mapOpacity: 0.12,
    },
    low: {
      label: 'LOW',
      color: '#16A34A',
      bgColor: 'rgba(22, 163, 74, 0.1)',
      borderColor: 'rgba(22, 163, 74, 0.3)',
      mapColor: '#16A34A',
      mapOpacity: 0.10,
    },
  };
  return meta[riskLevel] || meta.low;
}

/**
 * Get marker color based on verification status.
 */
export function getMarkerColor(verificationStatus) {
  const colors = {
    independent: '#16A34A',   // Green
    confirmed: '#2563EB',     // Blue
    unconfirmed: '#D97706',   // Amber
    similar: '#9CA3AF',       // Gray
    duplicate: '#6B7280',     // Muted gray
  };
  return colors[verificationStatus] || '#9CA3AF';
}

/**
 * Get marker opacity based on verification status.
 * Duplicates and unconfirmed are visually de-emphasized.
 */
export function getMarkerOpacity(verificationStatus) {
  const opacities = {
    independent: 1.0,
    confirmed: 0.9,
    unconfirmed: 0.6,
    similar: 0.4,
    duplicate: 0.35,
  };
  return opacities[verificationStatus] || 0.5;
}

/**
 * Calculate cluster radius in meters for map display.
 */
export function clusterRadiusMeters(radiusKm) {
  return radiusKm * 1000;
}

File: backend/utils/clustering.py
"""
HyperCrop — Clustering Utilities

Prototype clustering and risk classification logic.
Replace with DBSCAN, HDBSCAN, or a custom spatio-temporal
clustering algorithm for production deployments.
"""

import math
from .evidence import _haversine, classify_cluster_risk


def detect_clusters(reports: list, distance_threshold_km: float = 2.0) -> list:
    """
    Prototype cluster detection using simple distance-based grouping.

    Replace with DBSCAN or HDBSCAN with spatio-temporal features
    for production-grade clustering.

    Args:
        reports: List of report dicts with latitude, longitude, crop, suspectedIssue
        distance_threshold_km: Maximum distance to consider reports in the same cluster

    Returns:
        List of cluster dicts
    """
    if not reports:
        return []

    # Group by crop + issue first
    groups = {}
    for r in reports:
        key = f"{r['crop']}|{r['suspected_issue']}"
        if key not in groups:
            groups[key] = []
        groups[key].append(r)

    clusters = []
    cluster_id = 1

    for key, group_reports in groups.items():
        crop, issue = key.split('|')
        # Simple greedy clustering by distance
        assigned = set()

        for i, report in enumerate(group_reports):
            if i in assigned:
                continue

            cluster_members = [report]
            assigned.add(i)

            for j, other in enumerate(group_reports):
                if j in assigned:
                    continue
                dist = _haversine(
                    report['latitude'], report['longitude'],
                    other['latitude'], other['longitude']
                )
                if dist <= distance_threshold_km:
                    cluster_members.append(other)
                    assigned.add(j)

            if len(cluster_members) >= 2:
                cluster = _build_cluster(
                    f'CL-{cluster_id:02d}',
                    crop, issue, cluster_members
                )
                clusters.append(cluster)
                cluster_id += 1

    return clusters


def _build_cluster(cluster_id: str, crop: str, issue: str,
                   members: list) -> dict:
    """Build a cluster dict from its member reports."""
    lats = [m['latitude'] for m in members]
    lngs = [m['longitude'] for m in members]
    center_lat = sum(lats) / len(lats)
    center_lng = sum(lngs) / len(lngs)

    # Calculate radius
    max_dist = 0
    for m in members:
        d = _haversine(center_lat, center_lng, m['latitude'], m['longitude'])
        if d > max_dist:
            max_dist = d
    radius = round(max_dist + 0.1, 1)  # Add small buffer

    # Count by status
    independent = sum(1 for m in members
                      if m.get('verification_status') == 'independent')
    confirmed = sum(1 for m in members
                    if m.get('verification_status') in ('confirmed', 'independent'))
    unconfirmed = sum(1 for m in members
                      if m.get('verification_status') == 'unconfirmed')
    duplicate = sum(1 for m in members
                    if m.get('verification_status') in ('duplicate', 'similar'))

    scores = [m.get('evidence_score', 0.5) for m in members]
    avg_score = sum(scores) / len(scores) if scores else 0.5

    risk = classify_cluster_risk(independent, confirmed, len(members), avg_score)

    # Determine village from most common
    villages = [m.get('village', 'Unknown') for m in members]
    village = max(set(villages), key=villages.count)

    return {
        'id': cluster_id,
        'crop': crop,
        'issue': issue,
        'village': village,
        'center_latitude': round(center_lat, 4),
        'center_longitude': round(center_lng, 4),
        'radius_km': radius,
        'total_reports': len(members),
        'independent_reports': independent,
        'confirmed_reports': confirmed,
        'unconfirmed_reports': unconfirmed,
        'duplicate_reports': duplicate,
        'risk_level': risk,
        'avg_evidence_score': round(avg_score, 2),
    }
